### PR TITLE
docs: add page on query performance (close #2316)

### LINKED ISSUE #2316: Using variables destroys performance
Query without variables:
```
{
  user_news(where: {user_id: {_eq: 1}, status: {_eq: 0}}, limit: 10, order_by: {id: desc}) {
    id
  }
}
```
Throughput: **3800** req/s

Query with variables:
```
query($user:Int, $status: Int){
  user_news(where: {user_id: {_eq: $user}, status: {_eq: $status}}, limit: 10, order_by: {id: desc}) {
    id
  }
}
```
Variables:
```
{
  "user":1,
  "status":0
}
```
Throughput: **980** req/s

In more complex example where I first noticed this performance drop it went from ~1500req/s to ~250req/s.

### PR DESCRIPTION
**Pull request by marionschleifer**
### Description

Add a page about query performance including
- How to identify query runtime
- How query plan caching works in Hasura
- How to optimize performance

### Affected components
- Docs

### Changed pages:
- Added this page: https://deploy-preview-3693--hasura-docs.netlify.com/graphql/manual/queries/performance.html
- Added a reference here: https://deploy-preview-3693--hasura-docs.netlify.com/graphql/manual/queries/variables-aliases-fragments-directives.html#variables-and-performance

### CODE REVIEW

**Comment by marionschleifer** [comment]
@0x777 is there any additional information that we could add to this section? 🙂

File: docs/graphql/manual/queries/variables-aliases-fragments-directives.rst [hunk 1]
```diff
@@ -44,6 +44,14 @@ In order to make a query re-usable, it can be made dynamic by using variables.
       "author_id": 1
     }
 
+Variables and performance
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To optimise for performance, it's essential that the query consists of non-nullable variables only.
+Variables are non-nullable when they have a ``!`` after the type, e.g.: ``query getArticles($author_id: Int!)``.
```
Comment (rikinsk) [inline docs/graphql/manual/queries/variables-aliases-fragments-directives.rst]: Just `e.g.` in place of `e.g,:`. Currently looks a little akward.

File: docs/graphql/manual/queries/performance.rst [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+A plan cannot be generated for such queries because the variables defined in the GraphQL query don't have a one-to-one correspondence to the parameters in the prepared statement.
+
+Query optimization
+------------------
+
+Using GraphQL variables
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to leverage Hasura's query plan caching (as explained in the :ref:`previous section <query_plan_caching>`) to the full extent, GraphQL queries should be defined with
+variables whose types are **non-nullable** whenever possible.
```
Comment (0x777) [inline docs/graphql/manual/queries/performance.rst:167]: **non-nullable scalars**

File: docs/graphql/manual/queries/performance.rst [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+If the ``!`` is not added and the variable is nullable, the generated query will be different depending if an ``id`` is passed or if the variables is ``null``
+(for the latter, there is no ``where`` statement present). Therefore, it's not possible for Hasura to create a reusable plan for a query in this case.
+
+.. note::
+
+   Hasura is fast even for queries which cannot have a reusable plan.
+   This should concern you only if you face a high volume of traffic (thousands of requests per second).
+
+Using PG indexes
```
Comment (0x777) [inline docs/graphql/manual/queries/performance.rst:189]: Why not move this section to immediately after Analyze?

Comment (rikinsk) [reply docs/graphql/manual/queries/performance.rst:189]: I think the intent was to make sure someone reading the query optimisation section specifically is made aware of this fact. The earlier section is more of a reference that ppl might skip

File: docs/graphql/manual/queries/performance.rst (hunk 2 above)
Comment (rikinsk) [reply docs/graphql/manual/queries/performance.rst]: updated

**Pull request merged**